File: backup-db.js
const BashShell = require("./src/lib/BashShell.js");
const fs = require("fs");

const wait = ms => new Promise(resolve => setTimeout(resolve, ms));

const mainServer = new BashShell("BackupDB");
let busy = false;
mainServer.handler = function(event) {
    if (event.data) {
        console.log(event.data);
        if (event.data.includes("exported")) {
            busy = false;
        }
    }
};

async function exportCollection(collection) {
    const db = "vxsacademy";
    const outPath = `./db-backup/${db}.${collection}.json`;

    // use mongoexport to get data
    busy = true;
    mainServer.send(`mongoexport --collection=${collection} --db=${db} --out=${outPath}`);
    while (busy) {
        await wait(10);
    }
    
    // format data
    let writtenData = fs.readFileSync(outPath).toString().split("\n");
    for (let i = 0; i < writtenData.length; i++) {
        if (writtenData[i].length > 0) {
            let doc = JSON.parse(writtenData[i]);
            delete doc._id;
            writtenData[i] = JSON.stringify(doc);
        }
    }
    let newData = writtenData.join(",\n");
    newData = newData.slice(0, newData.length - 2);

    // write formatted data
    fs.writeFileSync(outPath, `[\n${newData}\n]`);
}

async function main() {
    await exportCollection("discussions");
    await exportCollection("programs");
    await exportCollection("salts");
    await exportCollection("users");
    mainServer.kill();
    console.log("Export Complete");
}

main();

File: utils/backup-db.dart
import 'dart:io';
import 'dart:convert';

void exportCollection(collection) {
    const db = "vxsacademy";
    final outPath = "./db-backup/${db}.${collection}.json";

    // use mongoexport to get data
    Process.runSync("mongoexport", "--collection=${collection} --db=${db} --out=${outPath}".split(" "));

    // format data
    var writtenData = File(outPath).readAsStringSync().split("\n");
    for (var i = 0; i < writtenData.length; i++) {
        if (writtenData[i].isNotEmpty) {
            Map doc = json.decode(writtenData[i]);
            doc.remove("_id");
            writtenData[i] = json.encode(doc);
        }
    }
    var newData = writtenData.join(",\n");
    newData = newData.substring(0, newData.length - 2);

    // write formatted data
    File(outPath).writeAsStringSync("[\n${newData}\n]");
}

void main() {
    exportCollection("discussions");
    exportCollection("programs");
    exportCollection("salts");
    exportCollection("users");
    print("Export Complete");
}
